plugins/chanbackup.c

```diff
@@ -0,0 +1,393 @@
+}
+
+static void update_scb(struct plugin *p, const char *scb_buf)
+{
+
+	/* If the temp file existed before, remove it */
+	unlink_noerr("scb.tmp");
+
+	int fd = open("scb.tmp", O_CREAT|O_EXCL|O_WRONLY, 0400);
+	if (fd<0)
+		plugin_err(p, "Opening: %s", strerror(errno));
+
+	plugin_log(p, LOG_DBG, "Updating the SCB file...");
+
+	write_hsm(p, fd, scb_buf);
+
+	/*~ fsync (mostly!) ensures that the file has reached the disk. */
+	if (fsync(fd) != 0) {
+		unlink_noerr("scb.tmp");
+	}
+
+	/*~ This should never fail if fsync succeeded.  But paranoia good, and
+	 * bugs exist. */
+	if (close(fd) != 0) {
+		unlink_noerr("scb.tmp");
+	}
+	/*~ We actually need to sync the *directory itself* to make sure the
+	 * file exists!  You're only allowed to open directories read-only in
+	 * modern Unix though. */
+	fd = open(".", O_RDONLY);
+	if (fd < 0) {
+		plugin_log(p, LOG_DBG, "Opening: %s", strerror(errno));
+	}
+	if (fsync(fd) != 0) {
+		unlink_noerr("scb.tmp");
+	}
+	close(fd);
+	unlink_noerr("emergency.recover");
+	rename("scb.tmp", "emergency.recover");
+}
+
+static struct command_result *after_staticbackup(struct command *cmd,
+					         const char *buf,
+					         const jsmntok_t *params,
+					         void *cb_arg UNUSED)
+{
+	const jsmntok_t *scbs = json_get_member(buf, params, "scb");
+	plugin_log(cmd->plugin, LOG_INFORM, "Updating the SCB");
+
+	update_scb(cmd->plugin, json_strdup(tmpctx, buf, scbs));
+	return notification_handled(cmd);
+}
+
+static struct command_result *json_state_changed(struct command *cmd,
+					         const char *buf,
+					         const jsmntok_t *params)
+{
+	const jsmntok_t *notiftok = json_get_member(buf,
+                                                    params,
+                                                    "channel_state_changed"),
+	*statetok = json_get_member(buf, notiftok, "new_state");
+
+	/* FIXME: I wanted to update the file on CHANNELD_AWAITING_LOCKIN,
+	 * But I don't get update for it, maybe because there is no previous_state,
+	 * also apparently `channel_opened` gets published when *peer* funded a channel with us?
+	 * So, is their no way to get a notif on CHANNELD_AWAITING_LOCKIN? */
+	if (json_tok_streq(buf, statetok, "CLOSED") ||
+		json_tok_streq(buf, statetok, "CHANNELD_NORMAL")) {
+
+		struct out_req *req;
+		req = jsonrpc_request_start(cmd->plugin,
+                                            cmd,
+                                            "staticbackup",
+                                            after_staticbackup,
+                                            &forward_error,
+                                            NULL);
+
+		return send_outreq(cmd->plugin, req);
+	}
+
+	return notification_handled(cmd);
+}
+
+static const char *init(struct plugin *p,
+			const char *buf UNUSED,
+			const jsmntok_t *config UNUSED)
+{
+	const char *scb_buf;
+	rpc_scan(p, "staticbackup",
+		 take(json_out_obj(NULL, NULL, NULL)),
+		 "{scb:%}", JSON_SCAN(json_to_str, &scb_buf));
```

Comment:
Hmm, converting to a string an back to JSON again is really ugly :(

Can you simply call staticbackup async here?  Maybe create a function to wrap this:

```
		struct out_req *req;
		req = jsonrpc_request_start(plugin, NULL,
                                            "staticbackup",
                                            after_staticbackup,
                                            &forward_error,
                                            NULL);

		return send_outreq(plugin, req);
```
?

Reply:
I Applied 200% of my brain and created `json_to_scb_chan`, which makes it easier for writing it in the file as well...